nemo/core/classes/common.py

```diff
@@ -317,20 +352,44 @@ def __call__(self, wrapped, instance: Typing, args, kwargs):
                 "not `input_ports() and `output_ports()`"
             )
 
+        # Preserve type information
+        if self.input_types is typecheck.TypeState.UNINITIALIZED:
+            self.input_types = instance.input_types
+
+        if self.output_types is typecheck.TypeState.UNINITIALIZED:
+            self.output_types = instance.output_types
+
+        # Resolve global type or local overridden type
+        if self.input_override:
+            input_types = self.input_types
+        else:
+            input_types = instance.input_types
+
+        if self.output_override:
+            output_types = self.output_types
+        else:
+            output_types = instance.output_types
+
+        # Inject type information into the instance of the class
+        # by using the wrapped function name. Can be used for introspection
+        wrapped_func_name = wrapped.__name__
+        setattr(instance, '__wrapped_fn_{}_input_types'.format(wrapped_func_name), input_types)
+        setattr(instance, '__wrapped_fn_{}_output_types'.format(wrapped_func_name), output_types)
```

Comment:
What are the reasons for this part?

Reply:
Due to dynamic change of type definition, external methods cannot introspect the wrapped functions actual type information. So say graphs require knowing the local function overriden information, they cannot obtain it from class scope. cls_obj.input_types and cls_obj.output_types will point to the class definition, not the local override.

So I am injecting the definition into the class during call so that after the call the graph can make multiple assumptions -
1) Function type is accessible by getattr - this will reflect either the cls_obj.*_type or the dynamically overriden local *_type.
2) As long as input and output types are not None, all properties of global typesystem will be available to local type overrides
3) If getattr fails, this function was never called.

I think these are valuable enough for the small cost of storing the dicts.

Reply:
After discussion, this has been reverted.

Reply:
I think graphs are/will operate on module instances, not classes. So in order to perform typecheck, binding etc.
I will simply access `obj.*_type`.

Reply:
`obj.*_type` <--- Is obj the output of the function (tensors with neural_type assigned)? Or the object of the class itself (obj = ClsWhichExtendsTyping())?

In the second case, this obj will no longer contain the actual overriden type anymore. It will only hold the global "class level" type (the default as defined by the user). Without the `

Reply:
updated comment: `obj.*_type` -> `obj.input_types()`/`obj.output_types()`

Reply:
`obj = ClsWhichExtendsTyping()`

Reply:
I understand that... but with my current understanding on "what do we want to achieve" I cannot even say whether it is good... or bad... or not important...

Reply:
Alright we can add it later if it's needed by graphs.